server/store/datastore/migration/017_gated_to_require_approval.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 Woodpecker Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migration
+
+import (
+	"fmt"
+
+	"src.techknowlogick.com/xormigrate"
+	"xorm.io/builder"
+	"xorm.io/xorm"
+)
+
+var gatedToRequireApproval = xormigrate.Migration{
+	ID: "gated-to-require-approval",
+	MigrateSession: func(sess *xorm.Session) (err error) {
+		const (
+			RequireApprovalNone         string = "none"
+			RequireApprovalForks        string = "forks"
+			RequireApprovalPullRequests string = "pull_requests"
+			RequireApprovalAllEvents    string = "all_events"
+		)
+
+		type repos struct {
+			ID              int64  `xorm:"pk autoincr 'id'"`
+			IsGated         bool   `xorm:"gated"`
+			RequireApproval string `xorm:"require_approval"`
+			Visibility      string `xorm:"varchar(10) 'visibility'"`
+		}
+
+		if err := sess.Sync(new(repos)); err != nil {
+			return fmt.Errorf("sync new models failed: %w", err)
+		}
+
+		// migrate gated repos
+		if _, err := sess.Exec(
+			builder.Update(builder.Eq{"require_approval": RequireApprovalAllEvents}).
+				From("repos").
+				Where(builder.Eq{"gated": true})); err != nil {
+			return err
+		}
+
+		// migrate public repos to new default require approval
+		if _, err := sess.Exec(
+			builder.Update(builder.Eq{"require_approval": RequireApprovalForks}).
```

Comment:
Should this stay as no approval at all to not risk breaking any setups?

Reply:
IMO it should. But if we decide to change the behavior in 2.8 already (which is also valid due to security reasons) we should take care of a clear note in the release notes and upgrade docs.

Reply:
I think having a changed default only in 3.0 makes things easier. So far, people also survived without this option and having in 3.0 is better WRT to semver.

Reply:
Changed it to be not breaking. Don't know however, how we would do the migration in 3.0 without overwriting explicit changes made in between.

Reply:
Add new type `NotSet = ""` and add smal new migration that set the new default in main ... + add smal if check for it and handle it acordingly ...

Reply:
We can also add a changelog note that gated field is depricated as of v2.8 so we can already remove it in v3.0.0 ...

Reply:
done with https://github.com/woodpecker-ci/woodpecker/pull/4429/commits/a7db4fe94e4d96a49e21f96327feee96dd1a1f4c
and https://github.com/woodpecker-ci/woodpecker/pull/4456